### PR TITLE
[LW] Resolve possible resource leak

### PR DESCRIPTION
**Pull request by Jolyon-S**
**Goals (and why)**:
Some amount of state is added during the start timestamp flow, but the relevant state is _not_ removed when a segment of the batch fails.

**Implementation Description (bullets)**:
* Add methods to remove state for failed start transaction attempts, and wire those through the start transaction batchers.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Multi namespace batched ones only

**Concerns (what feedback would you like?)**:
This fixes these issues, but not the general pattern. A followup PR will either make these weak/soft references, or just kill the cache store in general (although note that this does not make the changes in this PR irrelevant).

**Where should we start reviewing?**:
Diff

**Priority (whenever / two weeks / yesterday)**:
Sooner rather than later

### CODE REVIEW

File: lock-api/src/main/java/com/palantir/lock/client/BatchingCommitTimestampGetter.java [hunk 1]
```diff
@@ -73,7 +73,6 @@ public long getCommitTimestamp(long startTs, LockToken commitLocksToken) {
                 GetCommitTimestampsResponse response =
                         leaseService.getCommitTimestamps(requestedVersion, count - commitTimestamps.size());
                 commitTimestamps.addAll(process(batch.subList(commitTimestamps.size(), count), response, cache));
-                LockWatchLogUtility.logTransactionEvents(requestedVersion, response.getLockWatchUpdate());
```
Comment (Jolyon-S) [inline lock-api/src/main/java/com/palantir/lock/client/BatchingCommitTimestampGetter.java:76]: Drive-by

File: lock-api/src/test/java/com/palantir/lock/client/MultiClientTransactionStarterTest.java [hunk 2]
```diff
@@ -135,6 +137,7 @@ public void servesRequestsAsSoonAsResponseIsReceived() {
         LockCleanupService relevantLockCleanupService = LOCK_CLEANUP_SERVICE_MAP.get(namespace);
         verify(relevantLockCleanupService).refreshLockLeases(any());
         verify(relevantLockCleanupService).unlock(any());
+        verify(NAMESPACE_CACHE_MAP.get(namespace), times(4)).removeTransactionStateFromCache(anyLong());
```
Comment (Jolyon-S) [inline lock-api/src/test/java/com/palantir/lock/client/MultiClientTransactionStarterTest.java:140]: These numbers are magical to me; @sudiksha27 mind giving me some context so I can at least comment on why this is 4 (or 5 below)?

**Review by jeremyk-91 — APPROVED** [review]
should be good - but agree, would like comments on the magic 4 and 5

File: lock-api/src/test/java/com/palantir/lock/client/MultiClientTransactionStarterTest.java (hunk 2 above)
Comment (sudiksha27) [reply lock-api/src/test/java/com/palantir/lock/client/MultiClientTransactionStarterTest.java:140]: Right, so, for one namespace, we accumulate all requests for it in ResponseHandler#pendingFutures i.e. this queue is like [1, 25] for this test. Then we try to fetch as many startTs as possible from the server. For each batch of startTxnReponses, we try to server as many futures as we can. In one go our limit is 5 for these tests. So basically for one namespace i can fetch 5 responses.
If you see ResponseHandler#serveRequests, you will find, we greedily server requests from the pool of responses and the rest of the responses are transient i.e. they may be served if we have enough responses from future server calls. 
In this test, we can serve 1st request that needs 1 response. The next req needs 25 so we cannot serve it and therefore, there are 4 requests that need clearing up (server is configured to throw after one call)

File: lock-api/src/test/java/com/palantir/lock/client/MultiClientTransactionStarterTest.java [hunk 3]
```diff
@@ -166,6 +169,8 @@ public void shouldNotFreeResourcesIfRequestIsServed() {
         verify(LOCK_CLEANUP_SERVICE_MAP.get(alpha), never()).unlock(any());
         verify(LOCK_CLEANUP_SERVICE_MAP.get(beta)).refreshLockLeases(any());
         verify(LOCK_CLEANUP_SERVICE_MAP.get(beta)).unlock(any());
+        verify(NAMESPACE_CACHE_MAP.get(alpha), never()).removeTransactionStateFromCache(anyLong());
+        verify(NAMESPACE_CACHE_MAP.get(beta), times(5)).removeTransactionStateFromCache(anyLong());
     }
```
Comment (sudiksha27) [inline lock-api/src/test/java/com/palantir/lock/client/MultiClientTransactionStarterTest.java:174]: From above, the queue for beta here is [25] which means we cannot server any request. So the 5 responses we have from the first server call are transient and will be cleared.

**Review by sudiksha27 — COMMENTED** [review]
LGTM!
